### DISCUSSION TITLE
Native GUI builder

### DISCUSSION THREAD

**Discussion by Jonathanese** [Comfy-Org/ComfyUI#4361] [Ideas]
We have seen a lot of designs for new specific front-ends for ComfyUI in order to hide the spaghetti.

But there is a method of GUI building I would _love_ to see built into ComfyUI. I would design it myself, but I'm not familiar enough with the code yet.

To put it as simply as possible, I'll describe it as an action.

1. Go to a node like "Checkpoint Loader".
2. Right click on it
3. Under "Convert", you now have 3 options for each component: Input, Widget, GUI Element.
4. Set it to "GUI element"
5. The widget on the node itself doesn't disappear. However, in the (proposed) UI tab, the widget now appears there as well.

In the UI tab, you have a lock to switch between using the UI and editing the UI. When editing, you can drag the widgets around and arrange them how you want. You also have non-functional elements you can add like groups and labels.

Then that's pretty much it. Basically a page that could best be described as "show me absolutely nothing unless I explicitly add it to this page". And then for any component in the flow, you can just right click and add it to the UI panel.

Anything fancier like widget customization or something could be added in later. Or I might suggest adding the node's inputs/outputs to the node properties panel on the left, then having a drop down to select what each input/output is set to.

Too many attempts at ComfyUI frontends try to automate the whole process. They'll auto-detect any widgets and try to place them and name them automatically. They'll make all these assumptions about what you want displayed and where. But none of that is necessary. 

If we can just right click then add to the UI, you can get some impressive workflows with very simple interfaces. And these interfaces could be shared as part of the workflow data.

**Comment by ltdrdata** [comment]
Internally, discussions about various internal design improvements and roadmaps, including the proposed approach, are already taking place. Improvements to the ComfyUI front-end will be implemented gradually :)

**Reply by Jonathanese** [reply]
Great to hear!
